src/v1/album.js

```diff
@@ -51,46 +69,173 @@ router.put('/:id(\\w{24})', (req, res) => {
   }
 
   Album.getAlbum(req.params.id).then(album => {
-    album.name = req.body.name;
-    album.tracks = req.body.tracks || [];
-    album.images = req.body.images;
-
-    if (req.body.artist && req.body.artist != '') {
-      Artist.getArtist(req.body.artist)
-        .then(artist => {
-          var isExist = false;
-          artist.albums.forEach(obj => {
-            if (obj._id == req.body.artist) {
-              isExist = true;
-              return false;
-            }
-          });
+    if (!album) {
+      res.status(400)
+        .send(new Response.Error(new Err.ResourceNotFound(['album not found'])));
+    } else {
+      album.name = req.body.name;
+      album.images = req.body.images;
+      album.save().then(result => {
+        res.send(new Response.Data(album));
+      });
+    }
+  });
+});
 
-          if (!isExist) {
+router.post('/:id(\\w{24})/artist', (req, res) => {
+  if (!req.body.artist || req.body.artist == '') {
+    res.status(400)
+      .send(new Response.Error(new Err.InvalidParam(['artist is required'])));
+  } else {
+    Artist.getArtist(req.body.artist).then(artist => {
+      if (!artist) {
+        res.status(400)
+          .send(new Response.Error(new Err.ResourceNotFound(['artist not found'])));
+      } else {
+        Album.getAlbum(req.params.id).then(album => {
+          if (album.artist) {
+            res.status(400)
+              .send(new Response.Error(new Err.IllegalOperationError(['artist already set'])));
+          } else {
             artist.albums.push(album.toSimple());
-            artist.save();
+            album.artist = artist.toSimple();
+            Promise.all([artist.save(), album.save()]).then(result => {
+              res.send(new Response.Data(album));
+            });
           }
-          album.artist = artist.toSimple();
-          album.save();
-          res.status(200).send(new Response.Data(album));
         });
+      }
+    });
+  }
+});
+
+router.delete('/:id(\\w{24})/artist', (req, res) => {
+  Album.getAlbum(req.params.id).then(album => {
+    if (!album) {
+      res.status(400)
+        .send(new Response.Error(new Err.ResourceNotFound(['album not found'])));
     } else {
-      res.status(200).send(new Response.Data(album));
+      Artist.getArtist(album.artist._id.toHexString()).then(artist => {
+        _.remove(artist.albums, (_album) => {
+          return _album._id.equals(album._id);
+        });
+        album.artist = null;
+        Promise.all([artist.save(), album.save()]).then(result => {
+          res.send(new Response.Data(album));
+        });
+      });
     }
   });
 });
 
-router.delete('/:id(\\w{24})', (req, res) => {
-  Album.deleteAlbum(req.params.id)
-    .then(result => {
-      res.send(new Response.Data(result));
-    })
-    .catch(err => {
-      if (err instanceof Err.UnremovableError) {
-        res.status(400);
+router.put('/:albumId(\\w{24})/artist', (req, res) => {
```

Comment:
Is it possible to rename `albumId` to `id`? I'd like to make everything using the same rule.

Reply:
Fixed. =)